src/memory.rs

```diff
@@ -432,13 +392,13 @@ impl<'a, 'tcx> Memory<'a, 'tcx> {
     }
     
     // Returns a dtor and its argument, if one is supposed to run
-    pub(crate) fn fetch_tls_dtor(&mut self) -> Option<(ty::Instance<'tcx>, Pointer)> {
-        for (_, &mut TlsEntry { ref mut data, dtor }) in self.thread_local.iter_mut() {
-            if !data.is_null_ptr() {
-                if let Some(dtor) = dtor {
-                    let old_data = *data;
-                    *data = Pointer::from_int(0);
-                    return Some((dtor, old_data));
+    pub(crate) fn fetch_tls_dtor(&mut self) -> Option<(ty::Instance<'tcx>, PrimVal)> {
+        for (_, &mut TlsEntry { ref mut data, ref mut dtor }) in self.thread_local.iter_mut() {
+            if let Some(dtor) = dtor.take() {
```

Comment:
This changes the behavior, right? Now the dtor becomes `None`. If the dtor initializes this variable back to something non-NULL, this will mean the dtor is *not* run again. I think that's wrong, pthread says the dtors are run multiple times until the variables are all NULL.